### ISSUE TITLE
WpfPlot has slower framerate during mouse interactions in version 4.1

### ISSUE THREAD

**Issue by riquich** [ScottPlot/ScottPlot#761]
Hi, I admire and appreciate the effort put into this library and so I wanted to point out two things I noticed about 4.1 beta.

First is that the plots seems to stutter when panning in circular/cross directions, this might not be a major issue but I wanted to inform anyway since it might be root of other issues in the future. This problem is not present in the last public release (4.0.x).

Below I placed two gifs, with each containing signalplotconst containing 16k samples each. upper one is from 4.0.x and lower one is from 4.1.6. The stuttering is not present in 4.0.x. I tried to play with configuration of plot but it led to no avail. X axis labels have datetime formatting applied. I forgot to include rendering time but it cycles around 4-10ms. For bigger datasizes the stutters tend to occur more.

Second thing is that in version 4.1 rightclick event seems to be tied to automatic contextmenu generation  so it cannot be overridden. Even if it is overridden, the behavior given in lower gif occurs. I believe making rightclick event configurable would be better in terms of user point of view.

Gifs themselves might look sluggish due to my encoder but you can assume the movements are near-fluid.

Gif related to 4.0
![scottplot40](https://user-images.githubusercontent.com/34640902/107232266-4694d480-6a32-11eb-8f8f-713110638a25.gif)

Gif related to 4.1.6
![scottplot41](https://user-images.githubusercontent.com/34640902/107232414-6fb56500-6a32-11eb-84c6-b0d652f79a00.gif)

Thank you very much

**Comment by swharden** [comment]
Hi @riquich, thanks for the feedback!

## Stuttering in 4.1

This definitely warrants investigation and a fix! Is the stuttering you observe in the WinForms or WPF control?

## Right-click menu

Assuming this is WinForms, according to the demo application a custom right-click menu should be possible like this

https://github.com/ScottPlot/ScottPlot/blob/b22dca48e5e4c54600133a91dc7cf8736cb16215/src/demo/ScottPlot.Demo.WinForms/WinFormsDemos/RightClickMenu.cs#L20-L21

https://github.com/ScottPlot/ScottPlot/blob/b22dca48e5e4c54600133a91dc7cf8736cb16215/src/demo/ScottPlot.Demo.WinForms/WinFormsDemos/RightClickMenu.cs#L28-L34

Does this solve your right-click issue? I recognize this may not be easy to discover, but I'm working toward making a FAQ section and this will certainly be on it.

**Comment by riquich** [comment]
Hi @swharden sorry I forgot to state, this is observed in WPF. 

For the right click, I see the default event is public and it can be unsubscribed like you stated above, so this should be a solution thank you !

**Comment by riquich** [comment]
Also, i dont know if it matters but I am posting these additional info in any case for the behavior above:

Screen resolution: 2560x1440
DPI scaling: 125%
Framework: .NET Core 3.1

**Comment by swharden** [comment]
Thanks for the details @riquich!

The frame rate issue may relate to the "skip if currently rendering" functionality which seems implemented differently in the WinForms vs WPF control. Briefly, dropped frames should be permitted while dragging the mouse to keep the plot feeling responsive, so there's a chance this is where the slow frame rate is coming from.

https://github.com/ScottPlot/ScottPlot/blob/b22dca48e5e4c54600133a91dc7cf8736cb16215/src/controls/ScottPlot.WinForms/FormsPlot.cs#L56-L60

https://github.com/ScottPlot/ScottPlot/blob/b22dca48e5e4c54600133a91dc7cf8736cb16215/src/controls/ScottPlot.WPF/WpfPlot.xaml.cs#L64

**Comment by riquich** [comment]
At first I thought that might be the issue too, since skipIfCurrentlyRendering option is removed from Render function in 4.1, but now I checked that its default is set to false both in 4.0.x and 4.1.x Backends. For the code that produced 4.0.x gif above, the skipIfCurrentlyRendering parameter is not passed so it should be defaulted to false. If I am not missing something skip is only set to true in mousetracker and livegrowingdata examples so this issue seems to be stemming from somewhere else

**Comment by StendProg** [comment]
Hi @riquich , @swharden ,
The problem is in WPF debug mode.
Try it from `relese` or run from `exe` and it goes much more smother.
Debug session in WPF try to log every UI action (or something similar) and sometimes it lag's in non-obvious places.

**Comment by riquich** [comment]
Hi @StendProg I am aware of debug mode lag, it causes substantial random lag for both versions. However the gifs above are captured in release mode exes, and only 4.1.6 has stutter. By stutter i mean when mouse is down and panning in circular direction, you can see image is locked and is not being redrawn for 4.1.6 gif, which is not present in 4.0.x.

By the way near-horizontal or near-vertical only panning does not cause this issue

**Comment by StendProg** [comment]
>However the gifs above are captured in release mode exes, and only 4.1.6 has stutter.

This is very strange, I can easily reproduce the problem you described in debug mode. Even on a 50 point Scatter.
And the problem goes away completely if I run any `exe` 4.8 or 5.0. (40 ms `Signal Density` plot goes not so smoot, but never freezes like described above). 
I am using latest `master` branch.

**Comment by riquich** [comment]
It is strange indeed, I have been re-testing to see if I am doing something wrong but seeing the same results. But I have noticed one more thing, when the image gets "locked" as seen in the 4.1.6 gif, as long as i keep mouse moving, regardless of its direction or speed, as long as the mouse moves the plot never redraws, however when i stop moving the mouse it renders the image. The plots in the gifs are signalplotconst(maybe it might be the difference maker) with 16k samples(I tried 10mil samples and 1k samples but behavior is still the same). Thank you so much for your attention

**Comment by StendProg** [comment]
Hi @riquich,

I tried to reproduce this problem on another machine, with large datasets, and using the nuget `4.1.6-beta` package. And got the same result. The launch of individual `exe` works, albeit slowly, but there is no freezing.

In any case, the debugger mode is the main one for all of us, and we need to do something about it. The best solution (imho) would be to create a render queue using the approach used for rendering mouse wheel events. #545, #550.
I am ready to do it, but later and cannot say with certainty when, maybe within a week.

**Comment by riquich** [comment]
Hi @StendProg that might be somehow linked to a issue on my own setup then, thank you so much for your effort. This was not a major issue anyway. I will test on another computer when I get the chance to observe if the issue disappears. I am closing the issue for now then.

Again, thanks a lot

**Comment by swharden** [comment]
Thanks @StendProg for your input here! I'm glad you and @riquich were able to make some progress figuring this out.

> The best solution (imho) would be to create a render queue using the approach used for rendering mouse wheel events. #545, #550. 

@StendProg your mouse-wheel-zoom queue and event processor for anti-aliasing was very useful, and I am highly interested to see how much a render queue could improve performance if you decide to implement one. A non-blocking Render() method is also something I've been wanting to explore for a long time, and this may be a good way to get both features at the same time.

Also, it would be nice if there were a way for a low quality call to Render() to automatically perform a high quality render after 500ms. This functionality is usually achieved by mouse actions that trigger renders, but not when renders are called manually (e.g., when renders are requested from a method that is invoked when another plot's axis changes).

> I am ready to do it, but later and cannot say with certainty when, maybe within a week.

Take as long as you need! If it gets complicated or you want to discuss it further, we can start a new issue or discussion about it. I am very appreciative for the all work you put into this, and I am eager to see what you come up with.

**Comment by riquich** [comment]
I just wanted to post an update on this issue.

Yesterday I upgraded from 4.0.* to 4.1 and started seeing this issues again. So I started off to investigate what this problem's root cause is.

Started with a fresh project with only ScottPlot nuget installed. Tried all build configurations & framework types and result was: there was no plot locking (the problem i described above in gifs).

So I added my own libraries to this new project and tried again, still there was no plot locking.

Then I tried a different approach, I started plucking off one by one all of my UI elements from my original project to see if it has any effect and... viola. When I removed one specific UI element from the display the plot locking issue went away. After digging for some more I realized that object had a basic animation running (color of a tiny circle toggling between red and blue) and when I removed that animation the plot lock problem disappeared. I have to add a note though the color animated circle was hidden but seems like it still gets animated in hidden state.

One more thing: the animated object was there for both versions of ScottPlot 4.0 and 4.1 but only in 4.1 this causes an issue. And "UseRenderQueue" configuration prevents this locking but without it the plot interaction is more fluid. Writing this all so that if anyone else faces a similar problem they can have an idea. Anyway, finally sorted it all out. Thanks for discussion
